### PR TITLE
Create a tool like icepll for the ECP5 

### PR DESCRIPTION
**Pull request by C-Elegans**
Usage:
ecppll -i inputfreq -o outputfreq [-f file.v]

The PLLs this generates have not been tested on a device

### CODE REVIEW

**Comment by daveshah1** [comment]
Hi @C-Elegans !

Thank you for your great work on this! I am travelling back from CCC today but will do some testing on hardware once I'm home. 

One note is that Diamond does vary ICP_CURRENT and LPF_RESISTOR depending on the input configuration. I haven't investigated this fully yet, and I don't know whether these need to be set correctly for the PLL to work at all; or merely to optimise lock time and/or jitter.

In terms of the frequency calculation, I understand that the output divider is involved because the feedback path is via CLKOP. There are other feedback options available that could be investigated.

File: libtrellis/tools/ecppll.cpp [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+      }
+    }
+  }
+  return params;
+}
+
+void write_pll_config(pll_params params, const char* name,  FILE* file){
+  fprintf(file, "module %s(input clki, output clko);\n", name);
```
Comment (daveshah1) [inline libtrellis/tools/ecppll.cpp]: I don't have a problem merging this like this, but in general  the Trellis convention has been to use C++ libraries (fstream/iostream) over C.

**Comment by C-Elegans** [comment]
@daveshah1 I had noticed the LPF settings, and I assumed they had something to do with High vs. Low bandwidth mode, but after some testing, they also seem to vary with the input and output frequency. Generating the test data to figure this out will be a pain, because I have to have Clarity actually generate a PLL file, instead of just pressing the calculate button.

**Comment by C-Elegans** [comment]
There doesn't seem to be an easy to identify correlation between the LPF_Resistor and ICP current and any of input frequency, output frequency, PFD frequency, or VCO frequency.

[Data and graphs (PDF)](https://github.com/SymbiFlow/prjtrellis/files/2718728/lpf.and.icp.pdf)

**Comment by C-Elegans** [comment]
It looks to me that the values of LPF_RESISTOR and ICP_CURRENT are not critical for the PLL to work, and probably just affect the bandwidth or the amount of time it takes to lock. I tried synthesizing many different frequencies from the ecppll tool, from 20MHz to 400 MHz, and verified the frequencies on my oscilloscope, and it generated the correct frequency every time.

**Comment by daveshah1** [comment]
All sounds good, are you happy for me to merge this now?

**Comment by C-Elegans** [comment]
Yes please!

**Comment by daveshah1** [comment]
Thank you very much for your work on this!

**Pull request merged**